### PR TITLE
Peer attrition fix

### PR DESCRIPTION
**Pull request by Protryon**
This PR fixes issues observed with peer attrition.

### CODE REVIEW

File: network/src/peers/peer/connector.rs [hunk 1]
```diff
@@ -37,6 +37,14 @@ impl Peer {
...
             match self.inner_connect(node.version()).await {
                 Err(e) => {
+                    event_target
+                        .send(PeerEvent {
+                            address: self.address,
+                            data: PeerEventData::FailHandshake,
+                        })
+                        .await
+                        .ok();
+
```
Comment (ljedrz) [inline network/src/peers/peer/connector.rs:47]:
```suggestion
metrics::increment_gauge!(snarkos_metrics::queues::PEER_EVENTS, 1.0);
```

Comment (ljedrz) [reply network/src/peers/peer/connector.rs:47]: otherwise the related metric is off

File: network/tests/topology.rs [hunk 2]
```diff
@@ -42,6 +42,42 @@ async fn test_nodes<F: Fn() -> TestSetup>(n: usize, setup: F) -> Vec<Node> {
     nodes
 }
 
+#[tokio::test(flavor = "multi_thread")]
+async fn reconnect_nodes() {
+    // tracing_subscriber::fmt::init();
+    let setup = || TestSetup {
+        consensus_setup: None,
+        peer_sync_interval: 999999,
```
Comment (ljedrz) [inline network/tests/topology.rs:50]: I think the default is quite high, but no issue with it being bumped higher

**Review by ljedrz — APPROVED** [review]
LGTM with the metric adjustment in place :+1:

File: network/tests/topology.rs (hunk 2 above)
Comment (Protryon) [reply network/tests/topology.rs:50]: yeah, didn't want any peer syncs interfering with logic of the test.

**Comment by ljedrz** [comment]
Some extension to this fix might still be needed, as my node was still able to exhaust its peers and not seek further ones at all; the following is a `perf report` at the time when the node stopped doing anything:
```
Samples: 40  of event 'cycles:u', Event count (approx.): 245819
  Children      Self  Command          Shared Object       Symbol
+  100.00%     0.00%  tokio-runtime-w  libc-2.33.so        [.] __GI___clone (inlined)
+  100.00%     0.00%  tokio-runtime-w  libpthread-2.33.so  [.] start_thread
+  100.00%     0.00%  tokio-runtime-w  snarkos             [.] std::sys::unix::thread::Thread::new::thread_start
+  100.00%     0.00%  tokio-runtime-w  snarkos             [.] <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once (inlined)
+  100.00%     0.00%  tokio-runtime-w  snarkos             [.] core::ops::function::FnOnce::call_once{{vtable.shim}}
+  100.00%     0.00%  tokio-runtime-w  snarkos             [.] std::sys_common::backtrace::__rust_begin_short_backtrace
+  100.00%     0.00%  tokio-runtime-w  snarkos             [.] tokio::runtime::blocking::pool::Inner::run
+  100.00%     0.00%  tokio-runtime-w  snarkos             [.] tokio::runtime::task::harness::Harness<T,S>::poll
+  100.00%     0.00%  tokio-runtime-w  snarkos             [.] tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
+  100.00%     0.10%  tokio-runtime-w  snarkos             [.] tokio::runtime::thread_pool::worker::Context::run
+  100.00%     0.00%  tokio-runtime-w  snarkos             [.] tokio::runtime::thread_pool::worker::run
+  100.00%     0.00%  tokio-runtime-w  snarkos             [.] tokio::macros::scoped_tls::ScopedKey<T>::set
+   73.41%     0.00%  tokio-runtime-w  snarkos             [.] tokio::runtime::thread_pool::worker::Context::park_timeout
+   73.41%     0.00%  tokio-runtime-w  snarkos             [.] <tokio::runtime::park::Parker as tokio::park::Park>::park
+   55.44%     0.05%  tokio-runtime-w  snarkos             [.] tokio::time::driver::Driver<P>::park_internal
+   35.87%     0.00%  tokio-runtime-w  libc-2.33.so        [.] __clock_gettime_2 (inlined)
+   28.03%    28.03%  tokio-runtime-w  libc-2.33.so        [.] clock_gettime@GLIBC_2.2.5
+   19.77%    19.77%  tokio-runtime-w  snarkos             [.] tokio::time::driver::wheel::Wheel::poll
+   19.77%     0.00%  tokio-runtime-w  snarkos             [.] tokio::time::driver::<impl tokio::time::driver::handle::Handle>::process_at_time
+   18.66%    18.66%  tokio-runtime-w  snarkos             [.] tokio::runtime::queue::Steal<T>::steal_into
+   15.05%     0.00%  tokio-runtime-w  [unknown]           [.] 0xffffffff90a0007c
+   15.05%     0.00%  tokio-runtime-w  [unknown]           [.] 0xffffffff908a7281
+   12.77%     5.56%  tokio-runtime-w  snarkos             [.] parking_lot::condvar::Condvar::wait_until_internal
+    7.84%     7.84%  tokio-runtime-w  [unknown]           [k] 0xffffffff8fe70989
+    7.84%     0.00%  tokio-runtime-w  snarkos             [.] tokio::runtime::thread_pool::worker::Context::run_task
+    7.84%     0.00%  tokio-runtime-w  snarkos             [.] parking_lot::condvar::Condvar::notify_one_slow
+    7.84%     0.00%  tokio-runtime-w  [vdso]              [.] __vdso_clock_gettime
+    7.84%     0.00%  tokio-runtime-w  [unknown]           [.] 0xffffffff8ff36555
+    7.84%     0.00%  tokio-runtime-w  [unknown]           [.] 0xffffffff8ff35f45
+    7.84%     0.00%  tokio-runtime-w  [unknown]           [.] 0xffffffff8ff2d1fc
+    7.58%     0.06%  tokio-runtime-w  snarkos             [.] tokio::io::driver::Driver::turn
+    7.52%     0.00%  tokio-runtime-w  libc-2.33.so        [.] epoll_wait
+    7.52%     7.52%  tokio-runtime-w  [unknown]           [k] 0xffffffff90a00d70
+    7.22%     7.22%  tokio-runtime-w  [unknown]           [k] 0xffffffff8ff43720
+    7.22%     0.00%  tokio-runtime-w  libc-2.33.so        [.] syscall
+    5.19%     5.19%  tokio-runtime-w  libc-2.33.so        [.] __sched_yield
+    5.19%     0.00%  tokio-runtime-w  libc-2.33.so        [.] __GI___sched_yield (inlined)
     0.00%     0.00%  tokio-runtime-w  libc-2.33.so        [.] __libc_disable_asynccancel
```

**Comment by ljedrz** [comment]
Surprisingly, after around 30 minutes the nodes just decided to unblock itself and resumed regular work :eyes:; sadly, I missed the exact moment and don't know what happened at the very beginning, but I presume it was some disconnect that triggered it (it was at ~3 connections last time I checked before it fixed itself).

**Pull request merged**